### PR TITLE
Branches - Monick 

### PR DESCRIPTION
**Pull request by mfunkemomo**
# Heaps Practice

Congratulations! You're submitting your assignment!

## Comprehension Questions

Question | Answer
:------------- | :-------------
How is a Heap different from a Binary Search Tree? | Heap will ensure nodes are organized by least to greatest or greatest to least 
Could you build a heap with linked nodes? | Yes?
Why is adding a node to a heap an O(log n) operation? | Because you're adding it to only half the tree
Were the `heap_up` & `heap_down` methods useful?  Why? | Yes

### CODE REVIEW

File: lib/heap_sort.rb [hunk 1]
```diff
@@ -1,8 +1,34 @@
...
+    if sorted_list.empty? 
+      sorted_list.push(last_node)
+    elsif last_node < sorted_list[-1]
+      temp = sorted_list[-1]
+      sorted_list[-1] = last_node
+      sorted_list.push(temp)
+    else 
+      sorted_list.push(last_node)
+    end 
+  end
```
Comment (CheezItMan) [inline lib/heap_sort.rb:31]: This is a bit overly complicated.  Remember you always remove the smallest element of a heap.

```suggestion
  until heap.empty?
    sorted_list << heap.remove()
  end
```

File: lib/min_heap.rb [hunk 2]
```diff
@@ -14,18 +14,27 @@ def initialize
   end
 
   # This method adds a HeapNode instance to the heap
-  # Time Complexity: ?
-  # Space Complexity: ?
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def add(key, value = key)
```
Comment (CheezItMan) [inline lib/min_heap.rb:19]: 👍

File: lib/min_heap.rb [hunk 3]
```diff
@@ -14,18 +14,27 @@ def initialize
...
+    heap_up(@store.length - 1)
   end
 
   # This method removes and returns an element from the heap
   #   maintaining the heap structure
-  # Time Complexity: ?
-  # Space Complexity: ?
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def remove()
```
Comment (CheezItMan) [inline lib/min_heap.rb:29]: 👍

File: lib/min_heap.rb [hunk 4]
```diff
@@ -44,28 +53,55 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(n)
+  # Space complexity: O(1)
   def empty?
```
Comment (CheezItMan) [inline lib/min_heap.rb:58]: It works, but how is this O(n) time complexity?  Why not O(1)?

File: lib/min_heap.rb [hunk 5]
```diff
@@ -44,28 +53,55 @@ def to_s
...
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(log n)
+  # Space complexity: O(1)
   def heap_up(index)
```
Comment (CheezItMan) [inline lib/min_heap.rb:69]: Note, you're doing recursion so this is O(log n) space complexity.  You gotta remember the system stack!

File: lib/min_heap.rb [hunk 6]
```diff
@@ -44,28 +53,55 @@ def to_s
...
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    return if @store.empty?
+
+    left_child = 2*index+1
+    right_child = 2*index+2
+
+    # if either is nil, can use, so return
+    return if @store[left_child].nil? || @store[right_child].nil?
```
Comment (CheezItMan) [inline lib/min_heap.rb:92]: What if the left child is not `nil`, but it **IS** smaller than the parent?  Small edge-case the test missed here.

**Review by CheezItMan — COMMENTED** [review]
Overall pretty well done, some errors on BigO and one issue with heap_down.  See my inline comments and let me know what questions you have.  You did hit the main learning goals here.
